packages/flutter_driver/lib/src/driver.dart

```diff
@@ -0,0 +1,166 @@
+    new FlutterDriver._(serviceClient, isolate);
+
+// Package-private for internal tests.
+typedef Future<VMServiceClient> InternalConnectFunction(String url);
+
+// Package-private for internal tests.
+final InternalConnectFunction internalTrueConnectFunction = VMServiceClient.connect;
+
+// Package-private for internal tests.
+InternalConnectFunction internalConnectFunction = internalTrueConnectFunction;
+
+/// Drives a Flutter Application running in another process.
+class FlutterDriver {
+
+  static const String _flutterExtensionMethod = 'ext.flutter_driver';
+  static final Logger _log = new Logger('FlutterDriver');
+
+  /// Connects to a Flutter application.
+  ///
+  /// Resumes the application if it is currently paused (e.g. at a breakpoint).
+  ///
+  /// [dartVmServiceUrl] is the URL to Dart observatory (a.k.a. VM service). By
+  /// default it connects to `http://localhost:8181`.
+  static Future<FlutterDriver> connect({String dartVmServiceUrl: 'http://localhost:8181'}) async {
+    // Connect to Dart VM servcies
+    _log.info('Connecting to Flutter application at $dartVmServiceUrl');
+    VMServiceClient client = await internalConnectFunction(dartVmServiceUrl);
+    VM vm = await client.getVM();
+    _log.trace('Looking for the isolate');
+    VMIsolate isolate = await vm.isolates.first.load();
+    FlutterDriver driver = newFlutterDriver(client, isolate);
+
+    // Attempts to resume the isolate, but does not crash if it fails because
+    // the isolate is already resumed. There could be a race with other tools,
+    // such as a debugger, any of which could have resumed the isolate.
+    Future resumeLeniently() {
+      _log.trace('Attempting to resume isolate');
+      return isolate.resume().catchError((e) {
+        const vmMustBePausedCode = 101;
+        if (e is rpc.RpcException && e.code == vmMustBePausedCode) {
+          // No biggie; something else must have resumed the isolate
+          _log.warning(
+            'Attempted to resume an already resumed isolate. This may happen '
+            'when we lose a race with another tool (usually a debugger) that '
+            'is connected to the same isolate.');
+        } else {
+          // Failed to resume due to another reason. Fail hard.
+          throw e;
+        }
+      });
+    }
+
+    // Attempt to resume isolate if it was paused
+    if (isolate.pauseEvent is VMPauseStartEvent) {
+      _log.trace('Isolate is paused at start.');
+
+      // Waits for a signal from the VM service that the extension is registered
+      Future waitForServiceExtension() {
+        return isolate.onServiceExtensionAdded.firstWhere((VMServiceExtension ext) {
+          return ext.method == _flutterExtensionMethod;
+        });
+      }
+
+      // If the isolate is paused at the start, e.g. via the --start-paused
+      // option, then the VM service extension is not registered yet. Wait for
+      // it to be registered.
+      const timeout = const Object();
```

Comment:
`new Object()`.

We avoid `const Object()` because it's a bad pattern in some cases and we want to get out of the habit.